### ISSUE TITLE
Web API to return EF object with relationships not returning full result

### ISSUE THREAD

**Issue by ElvisHalil** [dotnet/efcore#5357]
When doing a chain of Includes and ThenIncludes to be returned by a Web API I only get a single record back with only the first relationship from the chain in the HTTP GET response. When I remove the whole chain and only leave the GetAll() method I get all the records back but without any relationships because I don't specify them any longer. 

Here is my method in my service class that I inject into my Web Api Controller, unitOfWork is my repository:

``` C#
        public IQueryable<EnforcementSession> GetAllSessions()
        {
            return _unitOfWork.SessionRepository.GetAll().AsQueryable()
                .Include(x => x.SessionOperator).ThenInclude(o => o.Operator)
                .Include(x => x.ZonesInSession).ThenInclude(z => z.Infractions)
                .Include(x => x.ZonesInSession)
                    .ThenInclude(y => y.UnitDevice)
                    .ThenInclude(y => y.Device).ThenInclude(y => y.DeviceModel)
                .Include(x => x.SessionDocument);
        }
```

Then I have my Web Api Controller simply calling that service and the method:

``` C#

        public ApiController(IQueueManager queueManager)
        {
            _queueManager = queueManager;
        }

        [HttpGet]
        public IQueryable<EnforcementSession> GetAllSessions()
        {
            return _queueManager.GetAllSessions();
        }
```

When I call this API with fiddler I get the following response:

```
[{"ID":"254ab8fb-bcda-495f-b46a-01d452f4bfa2",
"Automated":false,
"StartTime":"2016-03-14T11:30:02-06:00",
"EndTime":"2016-03-14T13:30:03-06:00",
"ContractTime":0,
"SessionOperator":[{"EnforcementSessionID":"254ab8fb-bcda-495f-b46a-01d452f4bfa2","OperatorID":"97bd7d28-46ef-4df2-8d46-f7c060674043"}]}]
```

As you can tell it returns only 1 record (session) and it includes only the SessionOperator relationship.  That's the first Include in my chain of includes/thenincludes.  Also there should be 254 records being returned not just 1.  

When I remove the chain of includes to this:

``` C#
        public IEnumerable<EnforcementSession> GetAllSessionsApi()
        {
            return _unitOfWork.SessionRepository.GetAll();
        }
```

and call that with my API:

``` C#
        [HttpGet]
        public IEnumerable<EnforcementSession> GetAllSessions()
        {
            return _queueManager.GetAllSessionsApi();
        }
```

I get the following response:

```
[{"ID":"254ab8fb-bcda-495f-b46a-01d452f4bfa2","Automated":false,"StartTime":"2016-03-14T11:30:02-06:00","EndTime":"2016-03-14T13:30:03-06:00","ContractTime":0,"SessionOperator":null,"SessionDocument":null,"ZonesInSession":null},{"ID":"c7976d95-d0a3-4838-bdcb-022cd7b21ef2","Automated":false,"StartTime":"2016-03-07T16:37:53-07:00","EndTime":"2016-03-07T18:07:40-07:00","ContractTime":0,"SessionOperator":null,"SessionDocument":null,"ZonesInSession":null},{"ID":"c938b59f-4b5b-4eb8-8a77-02577540911e","Automated":false,"StartTime":"2016-02-12T10:18:42-07:00","EndTime":"2016-02-12T11:18:05-07:00","ContractTime":0,"SessionOperator":null,"SessionDocument":null,"ZonesInSession":null},{"ID":"302ffa22-78a1-4e46-839a-04e113f36355","Automated":false,"StartTime":"2016-04-27T14:07:00-06:00","EndTime":"2016-04-27T15:07:03-06:00","ContractTime":0,"SessionOperator":null,"SessionDocument":null,"ZonesInSession":null},{"ID":"08b67fd6-c5a0-48b1-91e5-053bb7b71088","Automated":false,"StartTime":"2016-02-01T10:17:09-07:00","EndTime":"2016-02-01T11:17:09-07:00","ContractTime":0,"SessionOperator":null,"SessionDocument":null,"ZonesInSession":null},{"ID":"866a0f9d-ba1e-4714-9780-058f7026621a","Automated":false,"StartTime":"2016-03-29T15:46:24-06:00","EndTime":"2016-03-29T16:46:04-06:00","ContractTime":0,"SessionOperator":null,"SessionDocument":null,"ZonesInSession":null},{"ID":"370363b4-02b0-4d40-9403-07da59f921a8","Automated":false,"StartTime":"2016-04-08T13:58:50-06:00","EndTime":"2016-04-08T14:58:05-06:00","ContractTime":0,"SessionOperator":null,"SessionDocument":null,"ZonesInSession":null},{"ID":"9c22424c-f978-4337-99f0-0897af44955c","Automated":false,"StartTime":"2016-02-17T15:00:34-07:00","EndTime":"2016-02-17T16:30:04-07:00","ContractTime":0,"SessionOperator":null,"SessionDocument":null,"ZonesInSession":null},{"ID":"9ec5e0a7-0cf8-49e8-8ffb-099b43c6cf48","Automated":false,"StartTime":"2016-03-03T16:33:03-07:00","EndTime":"2016-03-03T16:52:11-07:00","ContractTime":0,"SessionOperator":null,"SessionDocument":null,"ZonesInSession":null},{"ID":"1fa63b64-4361-4b60-ba7b-0b82975c49c5","Automated":false,"StartTime":"2016-03-01T10:27:35-07:00","EndTime":"2016-03-01T11:30:44-07:00","ContractTime":0,"SessionOperator":null,"SessionDocument":null,"ZonesInSession":null}

and so on... (I had to shorten this, the response is the full list of 254 records.)
```

Obviously no relationships are being returned now because I've not specified them.  I'd like to be able to get all my relationships back with Web API.  Not sure if this is an issue or maybe there is a different approach I can take with this?  

I want to point out that I've tried to remove some of the includes and leave only the first 2, the response I get is the same as leaving the whole chain in. 

Looking forward to a response!  Thanks!

**Comment by divega** [comment]
@ElvisHalil I am not sure how the `IQueryable<EnforcementSession>` is handled by the Web API stack but from looking the symptoms I suspect the result enumeration is not being completely consumed. If I had a complete repro, I would probably try a few things to rule causes out, e.g.:

1. Force the evaluation of the query, e.g. by adding `ToList()` in:

 ``` C#
        [HttpGet]
        public IEnumerable<EnforcementSession> GetAllSessions()
        {
            return _queueManager.GetAllSessionsApi().ToList();
        }
```

2. Enabling or disabling MARS in the connection, e.g. with `MultipleActiveResultSets=True` or `MultipleActiveResultSets=False` in the database connection string.

Moreover, in order to help us produce a repro, could you provide more details:

1. How the application was created, e.g. what options did you choose for project templates, which versions of ASP.NET, Web API and EF you are using.
2. If possible a small model and query that repros the issue.

Of course, if you can point us to a project that repros the issue that should help too.

**Comment by ElvisHalil** [comment]
@divega I think my response is too large on that one.  I have made another web api using include and ThenInclude but with a ID parameter which returns relations for a single record instead of all my records and it seems to work fine.  

I am returning IActionResult for the Web Api method on that one.

I changed my Web Api method GetAllSessions() which I posted above to return IActionResult and in fiddler I do see all the results in RAW view with the correct relationships but in the JSON view it states that the input string was not in correct format. I think this is an unrelated problem however, and I think my initial problem is no longer valid. 

I'll go ahead and close this issue as I do believe everything is working correctly.
